core/dbt/task/debug.py

```diff
@@ -95,81 +115,140 @@ def path_info(self):
+                    "Profile should not be None if loading profile completed"
+                )
 
-        if self.any_failure:
-            fire_event(
-                DebugCmdResult(msg=red(f"{(pluralize(len(self.messages), 'check'))} failed:"))
+            adapter_version: str = self._read_adapter_version(
+                f"dbt.adapters.{adapter_type}.__version__"
             )
+            fire_event(DebugCmdOut(msg="adapter type: {}".format(adapter_type)))
+            fire_event(DebugCmdOut(msg="adapter version: {}".format(adapter_version)))
+
+        # Get project loaded to do additional checks
+        load_project_status: SubtaskStatus = self._load_project()
+
+        dependencies_statuses: List[SubtaskStatus] = []
+        if self.args.connection:
+            fire_event(DebugCmdOut(msg="Skipping steps before connection verification"))
         else:
-            fire_event(DebugCmdResult(msg=green("All checks passed!")))
+            # this job's status not logged since already accounted for in _load_* commands
+            self.test_configuration(load_profile_status.log_msg, load_project_status.log_msg)
+            dependencies_statuses = self.test_dependencies()
+
+        # Test connection
+        self.test_connection()
+
+        # Log messages from any fails
+        all_statuses: List[SubtaskStatus] = [
+            load_profile_status,
+            load_project_status,
+            *dependencies_statuses,
+        ]
+        all_failing_statuses: List[SubtaskStatus] = list(
+            filter(lambda status: status.run_status == RunStatus.FAIL, all_statuses)
+        )
 
-        for message in self.messages:
-            fire_event(DebugCmdResult(msg=f"{message}\n"))
+        failure_count: int = len(all_failing_statuses)
+        if failure_count > 0:
+            fire_event(DebugCmdResult(msg=red(f"{(pluralize(failure_count, 'check'))} failed:")))
+            for status in all_failing_statuses:
+                fire_event(DebugCmdResult(msg=f"{status.summary_message}\n"))
+            return DEBUG_RUN_STATUS["fail"]
+        else:
+            fire_event(DebugCmdResult(msg=green("All checks passed!")))
+            return DEBUG_RUN_STATUS["pass"]
 
-        return not self.any_failure
+    # ==============================
+    # Override for elsewhere in core
+    # ==============================
 
     def interpret_results(self, results):
         return results
 
-    def _load_project(self):
-        if not os.path.exists(self.project_path):
-            self.project_fail_details = FILE_NOT_FOUND
-            return red("ERROR not found")
-
-        renderer = DbtProjectYamlRenderer(self.profile, self.cli_vars)
+    # ===============
+    # Loading profile
+    # ===============
 
-        try:
-            self.project = Project.from_project_root(
-                self.project_dir,
-                renderer,
-                verify_version=self.args.VERSION_CHECK,
+    def _load_profile(self) -> SubtaskStatus:
```

Comment:
Is there some commonality with https://github.com/dbt-labs/dbt-core/blob/587bbcbf0def587c876cfd7ffe984889ac776e9f/core/dbt/config/runtime.py#L58 that we could extract into a common/shared function?

Reply:
There is some commonalty but I'm concerned about messing with that. I redid this function to get rid of its 3+ side effects. The function you're pointing at affects a lot more than just `debug` and seems to approach loading the profile a bit different, so I'm hesitant to bring them in alignment in this PR.

Also, I have a feeling others will ask us to avoid affecting more than we have to with the size/user changes received from this PR.